vm/src/obj/objstr.rs

```diff
@@ -16,13 +18,71 @@ extern crate unicode_segmentation;
 
 use self::unicode_segmentation::UnicodeSegmentation;
 
-impl FromPyObject for String {
-    fn typ(ctx: &PyContext) -> Option<PyObjectRef> {
-        Some(ctx.str_type())
+#[derive(Clone, Debug)]
+pub struct PyString {
+    // TODO: shouldn't be public
+    pub value: String,
+}
+
+impl PyString {
+    pub fn endswith(
+        zelf: PyRef<Self>,
+        suffix: PyRef<Self>,
+        start: OptArg<usize>,
+        end: OptArg<usize>,
+        _vm: &mut VirtualMachine,
+    ) -> bool {
+        let start = start.unwrap_or(0);
+        let end = end.unwrap_or(zelf.value.len());
+        zelf.value[start..end].ends_with(&suffix.value)
     }
 
-    fn from_pyobject(obj: PyObjectRef) -> PyResult<Self> {
-        Ok(get_value(&obj))
+    pub fn startswith(
+        zelf: PyRef<Self>,
+        prefix: PyRef<Self>,
+        start: OptArg<usize>,
+        end: OptArg<usize>,
+        _vm: &mut VirtualMachine,
```

Comment:
Is it necessary to move the vm to be the last argument? I prefer having it at the start - especially as there's a variable number of other arguments?

Reply:
My hope is to actually:

- not require it to be present when unneeded (like in this case)
- allow it to be in an arbitrary position when it _is_ needed

I haven't figured out how to accomplish those just yet, but the reason I moved the vm to the final position was so the payload could be the `self` parameter (or in the case of `PyRef`, to be easily changeable to take `self: PyRef<Self>` via `arbitrary-self-types` once it lands), so that it's natural to call them from rust, i.e. `PyString::endswith(<str>, <str>)` vs `<str>.endswith(<str>)`.

Reply:
I would also prefer the vm argument to be the first, or in this case the second parameter after self.

Reply:
Will do!